ocaml/xapi-cli-server/cli_operations.ml

```diff
@@ -4310,56 +4310,65 @@ let vm_migrate printer rpc session_id params =
-            try
-              List.find
-                (fun (_, h) ->
-                  h.API.host_hostname = x
-                  || h.API.host_name_label = x
-                  || h.API.host_uuid = x
+        let host =
+          let host_matches x self =
+            let hostname () = remote (Client.Host.get_hostname ~self) in
+            let uuid () = remote (Client.Host.get_uuid ~self) in
+            let name_label () = remote (Client.Host.get_name_label ~self) in
+            hostname () = x || uuid () = x || name_label () = x
+          in
+          let matches, fail_msg =
+            match List.assoc_opt "host" params with
+            | Some x ->
+                (host_matches x, Printf.sprintf "Failed to find host: %s" x)
+            | None ->
+                ( (fun _ -> true)
+                , Printf.sprintf "Failed to find a suitable host"
                 )
-                all
-            with Not_found ->
-              failwith (Printf.sprintf "Failed to find host: %s" x)
-          else
-            List.hd all
+          in
+          let all_hosts = remote Client.Host.get_all in
+          match List.find_opt matches all_hosts with
+          | Some host ->
+              host
+          | None ->
+              failwith fail_msg
         in
-        let network, network_record =
-          let all = Client.Network.get_all_records remote_rpc remote_session in
-          if List.mem_assoc "remote-network" params then
-            let x = List.assoc "remote-network" params in
-            try
-              List.find
-                (fun (_, net) ->
-                  net.API.network_bridge = x
-                  || net.API.network_name_label = x
-                  || net.API.network_uuid = x
-                )
-                all
-            with Not_found ->
-              failwith (Printf.sprintf "Failed to find network: %s" x)
-          else
-            let pifs = host_record.API.host_PIFs in
-            let management_pifs =
-              List.filter
-                (fun pif ->
-                  Client.PIF.get_management remote_rpc remote_session pif
-                )
-                pifs
-            in
-            if List.length management_pifs = 0 then
-              failwith
-                (Printf.sprintf "Could not find management PIF on host %s"
-                   host_record.API.host_uuid
-                ) ;
-            let pif = List.hd management_pifs in
-            let net = Client.PIF.get_network remote_rpc remote_session pif in
-            (net, Client.Network.get_record remote_rpc remote_session net)
+        let network =
+          let network_matches x self =
+            let bridge () = remote (Client.Network.get_bridge ~self) in
+            let uuid () = remote (Client.Network.get_uuid ~self) in
+            let name_label () = remote (Client.Network.get_name_label ~self) in
+            bridge () = x || uuid () = x || name_label () = x
```

Comment:
The number of PIFs fetches is reduced now by only fetching the ones present in the host, instead of all in the pool. This reduces the number of queries done when selecting the networks.